Plugins: Support changing plugin IDs (hardcoded)

ryantxu opened the pull request:
This adds an "alias" property to plugins so they can be renamed and still work.

Rather than modifying the registry -- this adds fallback checks that will iterate all plugins when the selected one is not found.

Test with https://github.com/grafana/grafana/pull/68859

Fixes https://github.com/grafana/grafana/issues/17186

ryantxu commented on pkg/plugins/manager/store/store.go:
```diff
@@ -36,6 +36,12 @@ func New(pluginRegistry registry.Service) *Service {
+			if v.Alias == pluginID {
+				return v, true
+			}
+		}
```
^^^ this is the key backend piece

wbrowne replied:
Rather than having this logic in both the store and client, it feels like this belongs in the central [registry](https://github.com/grafana/grafana/blob/main/pkg/plugins/manager/registry/in_memory.go).

ryantxu commented on public/app/features/plugins/importPanelPlugin.ts:
```diff
@@ -14,13 +14,16 @@ export function importPanelPlugin(id: string): Promise<PanelPlugin> {
   }
 
-  const meta = config.panels[id];
+  const meta = config.panels[id] || Object.values(config.panels).find((p) => p.alias === id);
```
^^ this is the key frontend piece

sdboyer commented:
Oooh, interesting. And really, really significant. Naming is hard, and allowing additional fields can have far-reaching effects.

i haven't read the code here yet, but i would suggest that if we're going to introduce any additional naming fields here, it should only be done with a clear plan wrt e.g. marketplace namespacing, and converting all currently-core plugins to be external plugins, and the name changes that will entail.

wbrowne reviewed (commented):
Related to @sdboyer's point -  I think this idea needs some time in the oven. A better step in my mind is to solve for this immediate case for pyroscope by keeping the alias in code rather than plugin.json if we want to do this short term.

I took it for a spin and noticed some issues when trying to use an "old" datasource
![Screenshot 2023-05-05 at 13 17 16](https://user-images.githubusercontent.com/1173299/236457189-ffdc99ed-98f3-4f95-9117-49aed772262c.png)

Also the logs there's logs that complain with
>ERROR[05-05|12:55:19] Could not find plugin definition for data source logger=context userId=1 orgId=1 uname=admin datasource_type=phlare

Do we expect requests for the old data source through the API to work? Otherwise
```
{"message":"Not found","messageId":"plugin.notRegistered","statusCode":404,"traceID":""}
```

ryantxu commented:
@wbrowne -- dooh, i added the fallback when looking up a single item.  :) Now it works here also:

<img width="762" alt="image" src="https://user-images.githubusercontent.com/705951/236503797-f6d19ebb-0fe0-4c76-baa2-edbfffff190c.png">

ryantxu commented on pkg/plugins/manager/loader/loader.go:
```diff
@@ -263,6 +263,14 @@ func (l *Loader) createPluginBase(pluginJSON plugins.JSONData, class plugins.Cla
 	}
 
+	// Hardcode alias features
+	switch pluginJSON.ID {
```
@wbrowne -- I removed the setting from the definition, and now set it manually here

aocenas replied:
Why is this set here? I thought the `alias` property in plugin.json made sense.

ryantxu replied:
@aocenas -- I agree, but as a first step @wbrowne suggests that we only aim to support this one plugin

tolzhabayev commented:
The idea of being able to transfer a plugin to someone else / change org for example is something that comes up from time to time but was never a priority for us due to this only happening about 2-3 times a year and usually resolved through plugin migrations / users switching to another plugin.

I am not sure this PR solves this problems and it feels like we are using some workarounds like using `type` field for `plugin-id`. We will review the PR more thoroughly but I would ask not to merge it until we have reached an agreement.

ryantxu committed:
squash

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu committed:
check alias when finding client

ryantxu committed:
update codeowners

ryantxu commented:
@aocenas -- updated with the tests we discussed.  Everything seems to work now 🎉 

@tolzhabayev -- I have no plans to merge without consent from @wbrowne 😜 

Yes this this comes up a few times a year and always feels silly that we don't have a pattern for it.  It is especially weird when it is our own products and we continue to have [dlopes7-appdynamics-datasource](https://grafana.com/grafana/plugins/dlopes7-appdynamics-datasource/).  

I think this approach gives us a pretty good option moving forward, but I agree there is no solution to supporting multiple names in the past.  but at least the future tends to become the past much quicker than we all would like

ryantxu commented:
>  it feels like we are using some workarounds like using type field for plugin-id

I wish it were so easy!  "type" == "plugin.id", and the names flip around all over 😬

andresmgot commented:
> > it feels like we are using some workarounds like using type field for plugin-id
> 
> I wish it were so easy! "type" == "plugin.id", and the names flip around all over grimacing

IMO this would be a blocker, misusing the `Type` to set it as the `ID`. Why not adding the field `Alias` to `DataSourceListItemDTO` (as done for `PanelDTO`)

ryantxu commented:
> misusing the Type to set it as the ID

this PR does not make any changes to the usage of type vs ID -- the existing form is VERY confusing, but this PR does not attempt to fix or change that

andresmgot commented:
> this PR does not make any changes to the usage of type vs ID -- the existing form is VERY confusing, but this PR does not attempt to fix or change that

ah, I see now that there are several occurrences of using the `Type` as `ID` in the code already. That's indeed confusing :/

wbrowne reviewed (commented):
Still gives me the heebie jeebies, but let's see if some more shaping helps that go away.

Given that we essentially have two potential identifiers for a plugin now, have you thought about what happens if one tries spoof another? Or slightly differently, if you try to install a plugin where the alias conflicts with an existing plugin?

ryantxu committed:
Update pkg/plugins/plugindef/plugindef.cue

Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu commented on pkg/api/datasources.go:
```diff
@@ -730,6 +731,15 @@ func (hs *HTTPServer) convertModelToDtos(ctx context.Context, ds *datasources.Da
 		ReadOnly:         ds.ReadOnly,
 	}
 
+	if hs.pluginStore != nil {
```
we could, but then we would also have to update most test fixtures 😬 -- we can tackle that if this approach seems reasonable

ryantxu committed:
rename the plugin

ryantxu commented:
> have you thought about what happens if one tries spoof another? Or slightly differently, if you try to install a plugin where the alias conflicts with an existing plugin?

The behavior in this PR is that if a plugin is registered with a given id , it will be used.  Only when nothing resolves to that id, does it look for things matching an alias.  If multiple things have the alias set 🤷 it will return the first it finds.

I think that is a reasonable pattern -- it gives us a good migration path, and we can eventually warn people in the UI if it happens.

----

re, managing the lookup in the registry.  Yes, i think that probably makes more sense.  The current approach waits until things fail to even check alias, but we could proactively keep a 2nd lookup table.  If that seems more reasonable.

ryantxu committed:
keep alias in registry

ryantxu commented on pkg/api/datasources.go:
```diff
@@ -730,6 +731,15 @@ func (hs *HTTPServer) convertModelToDtos(ctx context.Context, ds *datasources.Da
 
+	if hs.pluginStore != nil {
+		if plugin, exists := hs.pluginStore.Plugin(ctx, ds.Type); exists {
+			dto.TypeLogoUrl = plugin.Info.Logos.Small
```
weirdly... it does not have one 😛 

DataSource vs DataSourceListItemDTO

ryantxu committed:
do rename in json parsing

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu committed:
codeowners

ryantxu committed:
fix test

ryantxu committed:
fix test

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu committed:
update name

ryantxu committed:
phlare to pyroscope

ryantxu committed:
phlare to pyroscope

ryantxu committed:
wire

ryantxu committed:
frontend settigns

ryantxu commented:
@wbrowne -- I split this into two PRs so the 50+ changes from the phlare rename are easier to exclude from the changes.  See https://github.com/grafana/grafana/pull/68859

wbrowne reviewed (commented):
Thanks for splitting this!

The parts that irk me is the `Plugins()` on the registry which is the only part of the API that kind of ignores the alias bit, which makes sense but requires us have to the logic elsewhere (like `frontendsettings.go` for example). Then the datasource override stuff too is a bit tricky as it can be not obvious why you get a different ID back from the consumer perspective. There's some things we could do there to tidy this up later though so now I'll circulate amongst the team for some more 👀

ryantxu commented on pkg/plugins/plugins.go:
```diff
@@ -108,6 +108,7 @@ type JSONData struct {
 	ID           string       `json:"id"`
 	Type         Type         `json:"type"`
 	Name         string       `json:"name"`
+	Alias        string       `json:"alias,omitempty"`
```
since hardcoding is the hack... why move this to another level that will make it harder in the future?

IIUC, to remove the hardcoding hack, we should just need to add it to the cue schema

ryantxu replied:
For internal/hardcoded plugins sure -- it makes no difference.  But if we want this to be driven by the plugin configuration for external plugins eg (dlopes7-appdynamics-datasource, etc) where would it get set?

I honestly do not feel very strongly about any particular solution -- there are many many ways to skin this chicken -- just do not want to keep kicking the can down the road and never finding any approach.

ryantxu replied:
> believe it's part of plugin.json

Fundamentally I think it should be, having it hardcoded for phlare is just a strategy to gain some confidence right?  We still need a solution that allows future plugins to rebrand without having a clear path forward.

wbrowne replied:
100% agree - I'm just saying that we should have things in their rightful place _for now_ and when we have the confidence as you say, then we go all in. `JSONData` is our model for plugin.json and since we're hardcoding things for now, custom fields like that get put on the top level

andresmgot commented:
> The behavior in this PR is that if a plugin is registered with a given id , it will be used. Only when nothing resolves to that id, does it look for things matching an alias. If multiple things have the alias set shrug it will return the first it finds.
> 
> I think that is a reasonable pattern -- it gives us a good migration path, and we can eventually warn people in the UI if it happens.

Thinking about this, I don't see it as a reasonable pattern. For example, if you leave the id `phlare` available again, someone could create their own plugin, using `phlare` as the ID and impersonate the existing plugin. This can be a security concern. I think this should behave as a redirection, where the existing `id` of the plugin reamins "reserved" so no one can use it again (but points to the new ID).

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu commented:
@andresmgot i don't really understand/agree with this as a "security" issue -- either plugins live in a world where we sign them and audit what gets into the catalog or they don't.  In the case where we audit... YES we should make sure that alias and duplicates are used wisely!!!!  we should not let someone publish a "phlare" plugin (and already would not allow that)


> before we make the ball of .Type = plugin.ID bigger, I would try to fix that first, untangling that code.

😬  -- i think this has been on a TODO list for 4+ years, so um... 🤷 


@wbrowne -- I'm happy to move this out of JSONData if that is really the only blocker, but If there are other things (or another approach!) lets identify that first.

mckn commented:
> > The behavior in this PR is that if a plugin is registered with a given id , it will be used. Only when nothing resolves to that id, does it look for things matching an alias. If multiple things have the alias set shrug it will return the first it finds.
> > I think that is a reasonable pattern -- it gives us a good migration path, and we can eventually warn people in the UI if it happens.
> 
> Thinking about this, I don't see it as a reasonable pattern. For example, if you leave the id `phlare` available again, someone could create their own plugin, using `phlare` as the ID and impersonate the existing plugin. This can be a security concern. I think this should behave as a redirection, where the existing `id` of the plugin reamins "reserved" so no one can use it again (but points to the new ID).
> 
> Also, before we make the ball of `.Type = plugin.ID` bigger, I would try to fix that first, untangling that code.

To me, this makes sense and it would be good to also cover this in grafana.com api so if you try to fetch information about any of the aliases/plugin ids it will redirect you to the latest one being used.

marefr reviewed (commented):
Sorry for being a bit out of the loop. I think this is a reasonable approach for a first step towards supporting this. As my comments suggests we need to do follow up work to support plugin.json and possibly multiple aliases. While doing this, discuss possible security aspects and try to refactor things so that aliasing logic is limited to only a few places (only one place would be optimal) but for the sake of phlare lets try get this merged and follow up on next steps later.

wbrowne commented:
> @andresmgot i don't really understand/agree with this as a "security" issue -- either plugins live in a world where we sign them and audit what gets into the catalog or they don't. In the case where we audit... YES we should make sure that alias and duplicates are used wisely!!!! we should not let someone publish a "phlare" plugin (and already would not allow that)
> 
> > before we make the ball of .Type = plugin.ID bigger, I would try to fix that first, untangling that code.
> 
> 😬 -- i think this has been on a TODO list for 4+ years, so um... 🤷
> 
> @wbrowne -- I'm happy to move this out of JSONData if that is really the only blocker, but If there are other things (or another approach!) lets identify that first.

I think that'd be all from my side I think - this is the simplest thing that gets us going. I fail to see a more straight forward path so l think once that's patched up we're good 👍

renovate[bot] committed:
Update dependency esbuild to v0.17.19 (#61522)

* Update dependency esbuild to v0.17.18

* chore(npm): update all esbuild dependencies to use esbuild@0.17.18

* build(esbuild): update configs so esbuild-loader matches loader from file extension

* build(webpack): fix failed monaco editor loading due to require not existing

* chore(esbuild): bump to 0.17.19

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jack Westbrook <jack.westbrook@gmail.com>

dependabot[bot] committed:
Bump actions/stale from 6 to 8 (#65309)

Bumps [actions/stale](https://github.com/actions/stale) from 6 to 8.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v6...v8)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

linoman committed:
Reword JWT config claim (#69356)

Co-authored-by: ptz-xyt <104002657+ptz-xyt@users.noreply.github.com>

monteiro-renato committed:
Metrics: Update Help to mention active viewers (#63384)

domasx2 committed:
NodeGraph: Add "key" prop to context menu items (#69360)

add key to context menu items

ashharrison90 committed:
Nested folders: add skeleton scaffolding (#69310)

add skeleton scaffolding

domasx2 committed:
TempoQueryEDitor: Respect onRunQuery updates (#69362)

make tempo query editor respect onRunQuery changes

joshhunt committed:
Chore: Update codeowners for api-keys frontend (#69363)

make authnz own api-keys frontend

aocenas committed:
Pyroscope: Fix wrong defaults when importing query from different datasource (#69307)

LudoVio committed:
Fix plugin metric (#69368)

17billion committed:
Dashboards: Improve delete dashboard performance due to slow annotations query (#68544)

* Add index on dashboard_id column in annotation table

* Added org_id condition to the delete query for dashboard deletion.

ashharrison90 committed:
News: Expose config option to disable News feed (#69365)

* customize news feed

* remove url customisation

ryantxu committed:
Schema: Remove ServiceAccount kind (#69194)

remove service account kind

gabor committed:
logs: context: removed unused code (#69383)

joey-grafana committed:
Tracing: Upgrade tracing data source configuration editors (#68764)

* Add tracing config sub sections

* Export common sections and update divider in additional settings section

* Max width and margin bottom

* Add feature name to config link

* Update SpanBarSettings

* remove import

tskarhed committed:
i18n: Sync extraction with main (#69381)

Run extract on main

mgyongyosi committed:
AuthN: Make clientTokenRotation work when Grafana is accessible on a sub url (#69385)

Fix clientTokenRotation for auth-proxy

gabor committed:
logs: context: removed unused code (#69392)

Jguer committed:
Devenv: Update auth devenv versions (#69370)

update auth devenv versions

vtorosyan committed:
Docs: Rename External Group Sync references to Team Sync (#69395)

ktw4071 committed:
add datasource type (#69407)

lwandz13 committed:
update prometheus template variables doc (#69335)

* edits for clarity

* added 2 new sections on query variable options

* Update docs/sources/datasources/prometheus/template-variables/index.md

Co-authored-by: Brendan O'Handley <brendan.ohandley@grafana.com>

---------

Co-authored-by: Brendan O'Handley <brendan.ohandley@grafana.com>

svennergr committed:
Log Context: Fix split view button using the wrong query (#69369)

* fix wrong query used in split button

* refactor into one function

* don't act?

stephaniehingtgen committed:
Secure socks proxy: allow for UI to be disabled (#69408)

ryantxu commented on packages/grafana-data/src/types/plugin.ts:
```diff
@@ -58,6 +58,7 @@ export interface PluginMeta<T extends KeyValue = {}> {
   info: PluginMetaInfo;
   includes?: PluginInclude[];
   state?: PluginState;
+  alias?: string;
```
:) I started with an array, but made it singular so we don't blow anyones mind!

ryantxu committed:
moved alias field out of jsondata

ryantxu committed:
Merge remote-tracking branch 'origin/main' into plugin-alias-support

ryantxu commented:
@wbrowne -- I moved the field in https://github.com/grafana/grafana/pull/67867/commits/43e00c5a532e925ed2f07e2a9a3d25bd889a11e3 

also changed this so it is not "add to changelog" since there are no external implications
